src/python/pants/backend/shell/goals/test_test.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from pathlib import Path
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.shell.goals import test
+from pants.backend.shell.goals.test import ShellTestRequest, TestShellCommandFieldSet
+from pants.backend.shell.target_types import (
+    ShellCommandTarget,
+    ShellCommandTestTarget,
+    ShellSourcesGeneratorTarget,
+)
+from pants.build_graph.address import Address
+from pants.core.goals.test import TestResult, get_filtered_environment
+from pants.core.util_rules import archive, source_files
+from pants.engine.rules import QueryRule
+from pants.engine.target import Target
+from pants.testutil.rule_runner import RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    rule_runner = RuleRunner(
+        rules=[
+            *test.rules(),
+            *source_files.rules(),
+            *archive.rules(),
+            get_filtered_environment,
+            QueryRule(TestResult, (ShellTestRequest.Batch,)),
+        ],
+        target_types=[
+            ShellSourcesGeneratorTarget,
+            ShellCommandTarget,
+            ShellCommandTestTarget,
+        ],
+    )
+    rule_runner.set_options([], env_inherit={"PATH"})
+    return rule_runner
+
+
+def test_shell_command_as_test(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                shell_sources(name="src")
+
+                experimental_shell_command(
+                  name="msg-gen",
+                  command="echo message > msg.txt",
+                  tools=["echo"],
+                  outputs=["msg.txt"],
+                )
+
+                experimental_test_shell_command(
+                  name="pass",
+                  dependencies=[":msg-gen", ":src"],
+                  tools=["tr"],
```

Comment:
`tr` was used in a prior iteration but not the current iteration. However, the current shell backend code does not allow `tools` to be empty.